constraint/pkg/client/drivers/local/driver.go

```
@@ -0,0 +1,435 @@
+// a problem executing the query.
+func (d *Driver) eval(ctx context.Context, compiler *ast.Compiler, path []string, input ast.Value, opts ...drivers.QueryOpt) (rego.ResultSet, *string, error) {
+	cfg := &drivers.QueryCfg{}
+	for _, opt := range opts {
+		opt(cfg)
+	}
+
+	queryPath := strings.Builder{}
+	queryPath.WriteString("data")
+	for _, p := range path {
+		queryPath.WriteString(".")
+		queryPath.WriteString(p)
+	}
+
+	args := []func(*rego.Rego){
+		rego.Compiler(compiler),
+		rego.Store(d.storage),
+		rego.ParsedInput(input),
+		rego.Query(queryPath.String()),
+		rego.EnablePrintStatements(d.printEnabled),
+		rego.PrintHook(d.printHook),
+	}
+
+	buf := topdown.NewBufferTracer()
+	if d.traceEnabled || cfg.TracingEnabled {
+		args = append(args, rego.QueryTracer(buf))
+	}
+
+	r := rego.New(args...)
+	res, err := r.Eval(ctx)
+
+	var t *string
+	if d.traceEnabled || cfg.TracingEnabled {
+		b := &bytes.Buffer{}
+		topdown.PrettyTrace(b, *buf)
+		t = pointer.StringPtr(b.String())
+	}
+
+	return res, t, err
+}
+
+func (d *Driver) Query(ctx context.Context, target string, constraints []*unstructured.Unstructured, review interface{}, opts ...drivers.QueryOpt) ([]*types.Result, *string, error) {
+	if len(constraints) == 0 {
+		return nil, nil, nil
+	}
+
+	constraintsByKind := toConstraintsByKind(constraints)
+
+	traceBuilder := strings.Builder{}
+	constraintsMap := drivers.KeyMap(constraints)
+	path := []string{"hooks", "violation[result]"}
+
+	var results []*types.Result
+
+	// Round-trip review through JSON so that the review object is round-tripped
+	// once per call to Query instead of once per compiler.
+	reviewMap, err := toInterfaceMap(review)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for kind, kindConstraints := range constraintsByKind {
+		compiler := d.compilers.getCompiler(target, kind)
+		if compiler == nil {
+			// The Template was just removed, so ignore these Constraints.
```

Is this a state it's possible to be in? client acquires a read lock before it gathers the list of constraints.

If not, we may want to raise an error to make the unexpected state more visible

---

Modified to not allow concurrent changes to Client.

---

Done.